src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -927,7 +928,7 @@ private class DeclarationAnalysisData
         internal AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, Func<SyntaxNode, TLanguageKindEnum> getKind, AnalyzerManager analyzerManager) : base(analyzers, analyzerManager)
         {
             _getKind = getKind;
-            _declarationAnalysisDataCache = new ConditionalWeakTable<SyntaxReference, DeclarationAnalysisData>();
+            _declarationAnalysisDataCache = new ConditionalWeakTable<Compilation, Dictionary<Location, DeclarationAnalysisData>>();
```

Comment:
Probably just use a Dictionary as the value, and lock on it for operations. ConcurrentDictionary will also cause a similar contention when you try to fetch values from large number of threads.